cpp/oneapi/dal/backend/primitives/reducer.cpp

```diff
@@ -0,0 +1,221 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <algorithm>
+#include <cmath>
+#include <exception>
+#include <limits>
+
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+    #include <CL/sycl.hpp>
+#endif
+
+#include "oneapi/dal/backend/primitives/common.hpp"
+#include "oneapi/dal/backend/primitives/reducer.hpp"
+#include "oneapi/dal/data/array.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+
+namespace impl {
+
+template <typename UnaryFunctor, typename BinaryFunctor, typename Float, bool VectorsAreRows>
+struct reducer_singlepass_kernel {
+public:
+    const BinaryFunctor binary;
+    const UnaryFunctor unary;
+
+private:
+    typedef cl::sycl::
+        accessor<Float, 1, cl::sycl::access::mode::read_write, cl::sycl::access::target::local>
+            local_acc_t;
+
+public:
+    reducer_singlepass_kernel(BinaryFunctor binary_,
+                              UnaryFunctor unary_,
+                              std::int64_t vector_size_,
+                              size_t n_vectors_,
+                              const Float* vectors_,
+                              Float* reduces_,
+                              local_acc_t partial_reduces_)
+            : binary(binary_),
+              unary(unary_),
+              vector_size(vector_size_),
+              n_vectors(n_vectors_),
+              vectors(vectors_),
+              reduces(reduces_),
+              partial_reduces(partial_reduces_){};
+
+    void operator()(cl::sycl::nd_item<2> idx) {
+        const std::uint32_t local_size = idx.get_local_range(0);
+
+        const std::uint32_t global_dim = VectorsAreRows ? vector_size : 1;
+        const std::uint32_t local_dim  = VectorsAreRows ? 1 : n_vectors;
+
+        const std::uint32_t item_id  = idx.get_local_id(0);
+        const std::uint32_t group_id = idx.get_global_id(1);
+
+        Float el                 = vectors[group_id * global_dim + item_id * local_dim];
+        partial_reduces[item_id] = binary.init_value;
+
+        for (std::uint32_t i = item_id; i < vector_size; i += local_size) {
+            el                       = vectors[group_id * global_dim + i * local_dim];
+            partial_reduces[item_id] = binary(partial_reduces[item_id], unary(el));
```

Comment:
You could use a local variable instead of partial_reduces[item_id] which is placed in a local memory and write to the array when all calculations are done.